### PR TITLE
Add chacha20-poly1305 ciphersuites

### PR DESCRIPTION
**Pull request by mpg**
## Description

Implement CHACHA20-POLY1305 ciphersuites from RFC 7905.

## Status
**READY**

Note (just for information when planning the merge): this PR has non-trivial conflicts with #1633

## Requires Backporting

No, this is a new feature.

## Migrations

There is a default behaviour change, in that CHACHA20-POLY1305 ciphersuites are now preferred over AES-256-GCM suites, for performance and security in software implementations.

### CODE REVIEW

**Comment by mpg** [comment]
@sbutcher-arm I don't think there's much to review to design here, except the ciphersuite priority. This PR changes the preferred ciphersuite to a CHACHA20-POLY1305 suite, with the rationale that with software implementations (the ones that are provided with Mbed TLS), this has both better performance and fewer side-channel concerns than AES-GCM.

However, strictly speaking this is a change in observable behaviour: two Mbed TLS implementations with default settings communicating with each other will use a different ciphersuite after this PR than they did before (as evidenced by the need to adapt a test [here](https://github.com/ARMmbed/mbedtls/pull/1752/commits/ce66d5e8e1471bb60d726e4def0520f46e9c0057#diff-3c8a2535727055c7525b8ee52c73d896)). I think this is an acceptable change but I'd like your opinion on this.

Can you please review this design question and let me know what you think?

File: tests/compat.sh [hunk 1]
```diff
@@ -842,10 +865,17 @@ setup_arguments()
             exit 1;
     esac
 
+    # GnuTLS < 3.4 will choke if we try to allow CCM-8
+    if [ -z "${GNUTLS_MINOR_LT_FOUR-}" ]; then
```
Comment (AndrzejKurek) [inline tests/compat.sh]: Will this parameter expansion work in all shells that we support?

Comment (mpg) [reply tests/compat.sh]: It really should, as it's [prescribed by POSIX](http://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_06_02) and we already use it in `all.sh` (and the similar `:-` expansion in other places as well).

**Comment by mpg** [comment]
@hanno-arm This is a gentle reminder that I requested your review on this nice little 6-commit PR :)

@sbutcher-arm This is a gentle reminder that I requested your design review on one point of this PR (default preferred ciphersuite). In case you don't think you'll have time do it yourself, can you assign someone else to this design review? Thanks!

File: library/ssl_tls.c [hunk 2]
```diff
@@ -1394,62 +1408,84 @@ static int ssl_encrypt_buf( mbedtls_ssl_context *ssl )
...
+        {
+            /* GCM and CCM: concatenate fixed + explicit (=seqnum) */
+            memcpy( iv, transform->iv_enc, transform->fixed_ivlen );
+            memcpy( iv + transform->fixed_ivlen, ssl->out_ctr, 8 );
+            memcpy( ssl->out_iv, ssl->out_ctr, 8 );
+
+        }
+        else if( transform->ivlen == 12 && transform->fixed_ivlen == 12 )
+        {
+            /* ChachaPoly: XOR fixed + sequence number */
```
Comment (hanno-arm) [inline library/ssl_tls.c]: Nitpicking: I took the `+` literally at first, perhaps `ChaChaPoly: fixed XOR sequence number` is a bit easier to read.

File: library/ssl_tls.c [hunk 3]
```diff
@@ -1686,30 +1728,53 @@ static int ssl_decrypt_buf( mbedtls_ssl_context *ssl )
...
+        if( transform->ivlen == 12 && transform->fixed_ivlen == 4 )
+        {
+            /* GCM and CCM: concatenate fixed + explicit (transmitted) */
+            memcpy( iv, transform->iv_dec, transform->fixed_ivlen );
+            memcpy( iv + transform->fixed_ivlen, ssl->in_iv, 8 );
+
+        }
+        else if( transform->ivlen == 12 && transform->fixed_ivlen == 12 )
+        {
+            /* ChachaPoly: XOR fixed + sequence number */
```
Comment (hanno-arm) [inline library/ssl_tls.c]: Same comment as above.

**Comment by mpg** [comment]
@hanno-arm Thanks for the reviews! 
@AndrzejKurek I just pushed a new commit, can you review it as well?

@sbutcher-arm This PR now has two approving reviews and is only blocked by one design question (changing the default ciphersuite). Please either weigh in on this question or pass it on to someone else.

**Comment by sbutcher-arm** [comment]
I'm happy with the PR and the change, but I'd like @yanesca's approval of the change in the default behaviour to make Chacha-Poly the preferred ciphersuite, before I approve. 

@yanesca - Can you please approve that change? I can then give design approval.

**Pull request merged**